utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -708,27 +722,35 @@ void WriteBatchWithIndex::MultiGetFromBatchAndDB(
 
     // Note: we have to retrieve all columns if we have to merge KVs from the
     // batch and the DB; otherwise, the default column is sufficient.
+    // The columns field will be populated by the loop below to prevent issues
```

Comment:
Question for my own learning. I'm trying to understand how columns field being populated here would cause dangling pointer issue.

I was wondering why not just adding the `key_context` to `sorted_keys` below line 731 and 738 to avoid extra loop of `key_contexts` in line 744. We then won't preallocate size for `sorted_keys` like the other two, though.

Reply:
> Question for my own learning. I'm trying to understand how columns field being populated here would cause dangling pointer issue.

Right. So `std::vector` internally uses a contiguous heap-allocated buffer. As more and more elements are added to the vector, eventually we exceed the capacity of this buffer, at which point the vector has to allocate a bigger buffer (typical implementations double the capacity), and copy/move every item into the new buffer. This invalidates any pointer or iterator that points to the old buffer.

> 
> I was wondering why not just adding the `key_context` to `sorted_keys` below line 731 and 738 to avoid extra loop of `key_contexts` in line 744. We then won't preallocate size for `sorted_keys` like the other two, though.

It's actually another manifestation of the above issue; the internal buffer of `key_contexts` may get reallocated, which would render any pointer already added to `sorted_keys` invalid.

Reply:
Ah, so if the pointer to the tuple's `columns` (in the `merges` vector) is set in `key_context` here, then as soon as `merges` vector gets resized, that pointer is invalidated because the items are copied/moved to the new vector and the `key_context` no longer has the right value for the columns pointer. The same problem can happen if we make the `sorted_key` reallocated during population.

Please let me know if my understanding is correct. I think the change now makes perfect sense to me :)

Reply:
Yep, exactly!